Add ExtendRestAPI and SlotFills docs (#4208)
woocommerce/woocommerce-gutenberg-products-block: docs/extensibility/extend-rest-api-new-endpoint.md

# Adding an endpoint to ExtendRestAPI

> 

Reviewer: It might be worth including a small bit here at the top about when/why developers would want to make a new endpoint extensible.

Author: This document is mostly meant for contributors, I might rephrase it a bit to make that clear.

We usually decide as a team if we should add a new endpoint, based on developers feedback and needs, once that's settled, we go into the process of doing that.

Reviewer 2: ➕ 1 to making it more clear this documentation is for contributors to Woo Blocks.

> ## Exnteding `GET` endpoints in Store API.

Reviewer 3: ```suggestion
## Extending `GET` endpoints in Store API.
```

> ExtendRestAPI needs to expose to each endpoint Individually, meaning if you want to expose a new endpoint, you have to follow those steps:

Reviewer 3: ```suggestion
ExtendRestAPI needs to expose each endpoint individually. If you want to expose a new endpoint, you have to follow these steps:
```

> 1- In `ExtendRestApi` class, add your endpoint `IDENTIFIER` to the `$endpoints` variable.

Reviewer: I think you might want these steps to be formatted as `1. ` instead of `1- ` to render a bit more clearly in the doc:

```suggestion
1. In `ExtendRestApi` class, add your endpoint `IDENTIFIER` to the `$endpoints` variable.
```

```php

use Automattic\WooCommerce\Blocks\StoreApi\Schemas\BillingAddressSchema;

private $endpoints = [ /* other identifiers */, BillingAddressSchema::IDENTIFIER ];

```

This is to prevent accidentally exposing new endpoints.

> 2- Inside your endpoint schema class (for this example, inside `BillingAddressSchema`), in its `get_properties` method, add this call at the end of the returned array.

Reviewer: ```suggestion
2. Inside your endpoint schema class (for this example, inside `BillingAddressSchema`), in its `get_properties` method, add this call at the end of the returned array.
```

You can pass extra parameters to `get_extended_schema` and those would be passed to third party code.

```php
self::EXTENDING_KEY    => $this->get_extended_schema( self::IDENTIFIER ),
```

`EXTENDING_KEY` value is `extensions`, we use a constant to make sure we don't have a typo.

> 3- Inside the same class, in `get_item_response`, add the below line. Like `get_extended_schema`, you can pass extra parameters here as well.

Reviewer: ```suggestion
3. Inside the same class, in `get_item_response`, add the below line. Like `get_extended_schema`, you can pass extra parameters here as well.
```

Make sure to only expose what's needed.

```php
self::EXTENDING_KEY    => $this->get_extended_data( self::IDENTIFIER, $cart_item ),
```

That's it, your endpoint would now contain `extensions` in your endpoint, and you can consume it in the frontend.

> Extending a new endpoint is usally half the work, you will need to recvieve this data in frontend and pass it to any other extensiblity point (Slot, Filter, Event).

Reviewer 3: ```suggestion
Extending a new endpoint is usually half the work, you will need to receive this data in the frontend and pass it to any other extensibility point (Slot, Filter, Event).
```
